File: ViewAutomata/pda.cpp
#include "pda.h"

PDA::PDA(char epsilon, StatoPDA *s):epsilon(epsilon)
{
    partenza = s;
    stati = new std::list<StatoPDA*>();
}

StatoPDA *PDA::getStato(const std::string& s)
{
    std::list<StatoPDA*>::iterator i = stati->begin();
    while(i != stati->end() && (*i)->getNome() != s)
        i++;
    if (i == stati->end())
        return nullptr;
    else
        return (*i);
}

StatoPDA *PDA::operator[](const std::string& s)
{
    return getStato(s);
}

StatoPDA *PDA::operator[](int val)
{
    int count = 0;
    for(std::list<StatoPDA*>::iterator i = stati->begin(); i != stati->end(); i++)
    {
        if(count == val)
            return (*i);
        count++;
    }
    return nullptr;
}


int PDA::getNStati()
{
    return stati->size();
}

void PDA::addStato(const std::string& s, bool finale)
{
        if(getStato(s) == nullptr)
            stati->push_back(new StatoPDA(s, finale));
}

void PDA::removeStato(const std::string& s)
{
    bool eliminato = false;

    for(std::list<StatoPDA*>::iterator i = stati->begin();  i != stati->end(); i++)
    {

        for(int j = 0; j < (*i)->nTrans(); j++)
        {
            TransizionePDA* temp = static_cast<TransizionePDA*>((*i)->operator [](j));
            if(temp->getDest()->getNome() == s)
                (*i)->remove(dynamic_cast<StatoPDA*>(temp->getDest()),temp->getInput(),temp->getHead(),temp->getNewHead());
        }
    }
    for(std::list<StatoPDA*>::iterator i = stati->begin();  i != stati->end() && !eliminato; i++)
    {
        if((*i)->getNome() == s)
        {
            if(partenza->getNome() == s)
                partenza = nullptr;
            stati->erase(i);
            eliminato = true;
        }
    }
}

void PDA::setStartingState(const std::string& s)
{
    partenza = getStato(s);
}

PDA::~PDA()
{

}

void PDA::renameState(const std::string & oldName, const std::string & newName)
{
    StatoPDA* s = nullptr;
    bool trovato = false;
    for(std::list<StatoPDA*>::iterator i = stati->begin(); i != stati->end() && !trovato; i++)
    {
        if((*i)->getNome() == newName)
            trovato = true;
        else if((*i)->getNome() == oldName)
            s = (*i);
    }
    if(!trovato && s)
    {
        s->rename(newName);
    }
}

void PDA::addTransizione(const std::string & da, const std::string & a, char input, char head, std::string newHead)
{
    StatoPDA* daStato = getStato(da);
    StatoPDA* aStato = getStato(a);
    if (daStato && aStato)
        daStato->add(aStato,input,head,newHead);
}


char PDA::getEpsilon() const
{
    return epsilon;
}

void PDA::setEpsilon(char value)
{
    epsilon = value;
}

void PDA::supportChiusure(std::list<std::string> * curr, Stato* s)
{
    bool presente;
    for (int i = 0; i<s->nTrans(); i++)
    {
        if((*s)[i]->getInput() == epsilon)
        {
            presente = false;
            for(std::list<std::string>::iterator j = curr->begin(); j!=curr->end() && !presente; j++)
            {
                if(*j == (*s)[i]->getDest()->getNome())
                    presente = true;
            }
            if(!presente)
            {
                curr->push_back((*s)[i]->getDest()->getNome());
                supportChiusure(curr,(*s)[i]->getDest());
            }
        }
    }
}

void PDA::updateChiusure()
{
    std::list<std::string>* curr;
    chiusure->clear();
    for (std::list<StatoPDA*>::iterator i = stati->begin(); i != stati->end(); i++)
    {
        curr = new std::list<std::string>();
        curr->push_back((*i)->getNome());
        supportChiusure(curr,*i);
        chiusure->push_back(curr);
    }
}


bool PDA::check(StatoPDA * s, const std::string & input, const std::string& stack)
{
    std::string tempStack;
    TransizionePDA *tempTrans;
    std::list<StatoPDA*>* chiusuraCorr = new std::list<StatoPDA*>();
    for(std::list<std::list<std::string>*>::iterator i = chiusure->begin(); i != chiusure->end(); i++)
    {
        if((*i)->front() == s->getNome())
        {
            for(std::list<std::string>::iterator j = (*i)->begin(); j != (*i)->end(); j++)
            {
                chiusuraCorr->push_back(getStato(*j));
            }
        }
    }
    if(input.length() == 0)
    {
        if(stack == "")
            for(std::list<StatoPDA*>::iterator i = chiusuraCorr->begin(); i != chiusuraCorr->end(); i++)
                if((*i)->isFinale())
                    return true;
    }
    else
    {
        for(std::list<StatoPDA*>::iterator i = chiusuraCorr->begin(); i != chiusuraCorr->end(); i++)
        {
            for(int t = 0; t < (*i)->nTrans(); t++)
            {
                tempTrans = static_cast<TransizionePDA*>((*i)->operator [](t));
                if(tempTrans->getInput() == input[0] && (tempTrans->getHead() == stack[0] || tempTrans->getHead() == epsilon))
                {
                    tempStack = stack;
                    if(tempTrans->getHead() != epsilon)
                    {
                        tempStack = tempStack.substr(1,tempStack.length()-1);
                        if(tempTrans->getNewHead().length() != 0)
                        {
                            tempStack = tempTrans->getNewHead()[0] + tempStack;
                            if(tempTrans->getNewHead().length() == 2)
                                tempStack = tempTrans->getNewHead()[1] + tempStack;
                        }
                    }
                    else if(tempTrans->getNewHead().length() == 1)
                    {
                        tempStack = tempTrans->getNewHead()[0] + tempStack;
                    }
                    if(check(dynamic_cast<StatoPDA*>(tempTrans->getDest()),input.substr(1,input.length()-1),tempStack))
                        return true;
                    //delete tempTrans;
                }
            }
        }
    }
    return false;
}

bool PDA::start(const std::string & input)
{

    updateChiusure();
    if(partenza)
        return check(dynamic_cast<StatoPDA*>(partenza), input, "");
    else
        return false;
}

File: progettoP2/src/progettoP2/PDA.java
package progettoP2;

import java.util.ArrayList;
import java.util.Iterator;

public class PDA implements AbstractAutoma {
	private ArrayList<StatoPDA> stati;
    private StatoPDA partenza; 
    private ArrayList<ArrayList<String>> chiusure;
    private char epsilon; 
    
	protected PDA(){
		stati = new ArrayList<StatoPDA>();
		chiusure = new ArrayList<ArrayList<String>>();
	}
	protected PDA(StatoPDA s, char e){
    	partenza = s;
    	stati = new ArrayList<StatoPDA>();
    	chiusure = new ArrayList<ArrayList<String>>();
    	epsilon = e;
    }
	
	@Override
	public Boolean start(String s) {		
		updateChiusure();
	    if(partenza != null)
	        return check(partenza, s, "");
	    else
	        return false;
	}
	
	private void updateChiusure(){
		ArrayList<String> curr;
	    chiusure.clear();
	    for (StatoPDA temp : stati)
	    {
	        curr = new ArrayList<String>();
			curr.add(temp.getNome());
	        supportChiusure(curr,temp);
	        chiusure.add(curr);
	    }
	}
	
	private void supportChiusure(ArrayList<String> curr , Stato s ){
	    Boolean presente;
	    for (int i = 0; i<s.nTrans(); i++)
	    {
	        if(s.operatorBarraBarra(i).getInput() == epsilon)
	        {
	            presente = false;
	            for(int j = 0; j < curr.size() && !presente; j++)
	            {
	                if(curr.get(j) == s.operatorBarraBarra(i).getDest().getNome())
	                    presente = true;
	            }
	            if(!presente)
	            {
	                curr.add(s.operatorBarraBarra(i).getDest().getNome());
	                supportChiusure(curr,s.operatorBarraBarra(i).getDest());
	            }
	        }
	    }
	}
	
	private Boolean check(StatoPDA s, final String input, final String stack){
	    String tempStack;
	    TransizionePDA tempTrans;
	    ArrayList<StatoPDA> chiusuraCorr = new ArrayList<StatoPDA>();
	    for(int i = 0; i < chiusure.size(); i++)
	    {
	        if(chiusure.get(i).get(0) == s.getNome())
	        {
	            for(int j = 0; j < chiusure.get(i).size(); j++)
	            {
	                chiusuraCorr.add(getStato(chiusure.get(i).get(j)));
	            }
	        }
	    }
	    if(input.length() == 0)
	    {
	        if(stack.length() == 0)
	            for(int i = 0; i < chiusuraCorr.size(); i++)
	                if(chiusuraCorr.get(i).isFinale())
	                    return true;
	        
	        for(int i = 0; i < chiusuraCorr.size(); i++)
	        {
	            for(int t = 0; t < chiusuraCorr.get(i).nTrans(); t++)
	            {
	                tempTrans = (TransizionePDA) chiusuraCorr.get(i).operatorBarraBarra(t);
	                
	                if(tempTrans.getInput() == epsilon
	                		&& (stack.length() != 0 && (tempTrans.getHead()) == stack.charAt(0)))
	                {
	                    tempStack = stack;
	                    if(tempTrans.getHead() != epsilon)
	                    {
	                        tempStack = tempStack.substring(1);
	                        if(tempTrans.getNewHead().charAt(0) != epsilon)
	                        {
	                            tempStack = tempTrans.getNewHead() + tempStack;
	                        }
	                    }
	                    else 
	                    {	         
	                    	if(tempTrans.getNewHead().charAt(0) != epsilon) 
	                        {
	                    		tempStack = tempTrans.getNewHead().charAt(0) + tempStack;
	                        }
	                    }
	                    if(check((StatoPDA) tempTrans.getDest(),
	                    		tempTrans.getInput()==epsilon?
	                    		input:   
	                    		input.substring(1), tempStack))
	                        return true;
	                }
	            }
	        }
	    }
	    else
	    {
	    	   for(int i = 0; i < chiusuraCorr.size(); i++)
		        { 
		            for(int t = 0; t < chiusuraCorr.get(i).nTrans(); t++)
		            {
		                tempTrans = (TransizionePDA) chiusuraCorr.get(i).operatorBarraBarra(t);
		                
	                if((tempTrans.getInput() == input.charAt(0) &&  stack.length()!= 0 && tempTrans.getHead() == stack.charAt(0)) 
	                		|| tempTrans.getHead() == epsilon)
	                {
	                    tempStack = stack;
	                    if(tempTrans.getHead() != epsilon)
	                    {
	                        tempStack = tempStack.substring(1);
	                        if(tempTrans.getNewHead().charAt(0) != epsilon) // !!!!!
	                        {
	                            tempStack = tempTrans.getNewHead() + tempStack;
	                        }
	                    }
	                    else if(tempTrans.getNewHead().length() == 1)
	                    {
	                        if(tempTrans.getNewHead().charAt(0) != epsilon) // !!!!!
	                        {
	                            tempStack = tempTrans.getNewHead().charAt(0) + tempStack;
	                        }
	                    }
	                    if(check((StatoPDA)(tempTrans.getDest()),
	                    		tempTrans.getInput()==epsilon? input : input.substring(1),tempStack))
	                        return true;
	                }
	            }
	        }
	    }
	    return false;
	}
	
	@Override
	public void addState(String s) {
		if(getStato(s) == null)
            stati.add(new StatoPDA(s, false));
	}

	@Override
	public void removeState(String s) {
		Boolean eliminato = false;
	    for(int i = 0; i<stati.size() && !eliminato; i++)
	    {
	    	for(int j = 0; j<stati.get(i).nTrans(); j++){
	    		 TransizionePDA temp = (TransizionePDA)(stati.get(i).operatorBarraBarra(j));
	             if(temp.getDest().getNome() == s)
	                 stati.get(i).remove((StatoPDA)(temp.getDest()),temp.getInput(),temp.getHead(),temp.getNewHead());
	    	}
	    }
	    for(int i= 0; i<stati.size(); i++){
	    	if(stati.get(i).getNome() == s){
	    		if(partenza.getNome() == s){
	    			partenza = null;
	    		}
	    		stati.remove(i);
	    		eliminato = true;
	    	}
	    }
	}

	@Override
	public void setStartingState(String s) {
		partenza = getStato(s);
	}
	
	public StatoPDA getStato(final String s){	
		for(StatoPDA tmp : stati){
			if(s == tmp.getNome()){
				return tmp;
			}
		}
		return null;		 
	}
	
	public StatoPDA operatorBarraBarra(final String s){
		return getStato(s);
	}
	
	public StatoPDA operatorBarraBarra(final int ind){
		int i = 0;
		for(StatoPDA s : stati){
			if(i == ind){
				return s;
			}
			i++;
		}
		return null;
	}
	
	public int getNStati(){
		return stati.size();
	}
	
	@Override
	public void renameState(final String oldName, final String newName) {
		StatoPDA s = null;
	    Boolean trovato = false;
	    for(int i = 0 ; i < stati.size() && !trovato; i++)
	    {
	        if(stati.get(i).getNome() == newName){
	            trovato = true;
	        }
	        else if(stati.get(i).getNome() == oldName){
	            s = stati.get(i);
	        }
	    }
	    if(!trovato && s != null)
	    {
	        s.rename(newName);
	    }
		
	}
	
	public void addTransizione(final String da,
			final String  a,
			char input,
			char head, 
			String newHead)
	{
	    StatoPDA daStato = getStato(da);
	    StatoPDA aStato = getStato(a);
	    if (daStato != null && aStato != null)
	        daStato.add(aStato,input,head,newHead);
	}
	@Override
	public Stato getStartingState() {
		return partenza;
		
	}
	public char getEpsilon() {
		return epsilon;
	}
	public void setEpsilon(char epsilon) {
		this.epsilon = epsilon;
	}

}
